File: android/fundamental-core/src/test/kotlin/com/fundamental/core/AbsorbRSpecTests.kt
package com.fundamental.core

import com.fundamental.core.engine.Box
import com.fundamental.core.math.Vec3
import com.fundamental.core.runtime.BodySpec
import com.fundamental.core.runtime.createField
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * #1177 — `absorbR` had no path through `BodySpec` on either native plane, so a consumer holding a
 * BodyHandle was stuck at the Body default. That radius *is* the behaviour for `sink` (what it
 * captures) and sizes where `warp` ejects, and the engine's own tests reached through
 * `handle.body.absorbR` to work around it — a test proving something the public API could not express.
 */
class AbsorbRSpecTests {
    private fun box() = Box(center = Vec3(400f, 300f, 0f), halfExtents = Vec3(20f, 20f, 0f))

    @Test
    fun absorbRIsSettableThroughBodySpec() {
        val f = createField(800f, 600f, particleCount = 50, seed = 7)
        val h = f.addBody(BodySpec(tokens = listOf("sink"), absorbR = 90f, rect = { box() }))
        assertEquals(90f, h.body.absorbR, "the spec value must reach the body")
    }

    @Test
    fun omittingItKeepsTheBodyDefault() {
        val f = createField(800f, 600f, particleCount = 50, seed = 7)
        val h = f.addBody(BodySpec(tokens = listOf("sink"), rect = { box() }))
        assertEquals(10f, h.body.absorbR, "default must be unchanged by this addition")
    }
}

File: swift/Tests/FundamentalVanillaTests/AbsorbRSpecTests.swift
import Foundation
import Testing
@testable import FundamentalVanilla
import FundamentalCore

// #1177 — `absorbR` had no path through `BodySpec` on any plane, so a consumer holding a BodyHandle
// was stuck at the `Body` default. That radius *is* the behaviour for `sink` (the horizon it absorbs
// within) and for `warp` (the throat it transports through), so a programmatic body could not express
// what markup could. This mirrors the JS and Kotlin additions; the defaults are deliberately untouched.

@Suite("absorbR through BodySpec")
struct AbsorbRSpecTests {
    private func box() -> Box { Box(center: Vec3(400, 300, 0), halfExtents: Vec3(20, 20, 0)) }

    @Test("a spec value reaches the backing body")
    func specValueReachesBody() {
        let field = FieldField(host: HeadlessFieldHost())
        let h = field.addBody(BodySpec(tokens: ["sink"], absorbR: 90, rect: { self.box() }))
        #expect((h.bodyRef() as? Body)?.absorbR == 90)
        field.destroy()
    }

    @Test("omitting it leaves the body default alone")
    func omissionKeepsDefault() {
        let field = FieldField(host: HeadlessFieldHost())
        let a = field.addBody(BodySpec(tokens: ["sink"], rect: { self.box() }))
        // The `Body` default (Types.swift) — this addition is strictly additive and must not move it.
        #expect((a.bodyRef() as? Body)?.absorbR == 10)
        field.destroy()
    }
}
